### ISSUE TITLE
Shipit Revision check improvements 

### ISSUE THREAD

**Issue by byroot** [Shopify/shipit-engine#490]
### Context

Shipit is pretty much oblivious to what is running in production.
It was built on the idea that a revision is deployed or not.
When a deploy is triggered, if the deploy script crash, Shipit consider that nothing changed on our production servers.

The only feature we had, was to poll every minutes for the production revision, so that Shipit could be aware of manual deploys and reflect theses in it's data model.

But even this is based on the idea that deploys are atomic, and that a single revision is running at any moment.

Unfortunately this isn't true.

### Early work

https://github.com/Shopify/shipit-engine/pull/489 introduce a `flapping` status in Shipit, to reflect the situations were we're running 2 revisions.

The status itself make sense, but the way to detect it doesn't really. It reacts to inconsistent responses from the script. e.g:

- Deploy fail
- Revision check says the new revision is running
- Once the the check return the same result 5 times in a row, it is considered as the new truth.

This is a bit silly, the script should be able to tell shipit that 2 revisions are running immediately.

#### Manual deploy / rollback

This use case should still be supported. If someone deployed the stack manually, Shipit should be able to detect it to prevent any confusion.

#### Failed non-transactionnal deploy

The deploy script crash leaving nodes on 2 different revisions. Shipit need to mark the deploy as `flapping` until the situation is resolved.
Once we're back on a single revision, Shipit should mark this deploy as either `success` or `failure` based on which revision is finally in production.

### Proposal

We could have the script return a specific error code if multiple revisions are running:

  - `0`: A single revision is running. That revision is printed on STDOUT.
  - `42`: Multiple revisions are running. No specific output is expected.

Non-zero exits other than `42` are ignored. This is so that application with transactional deploys (e.g. heroku apps) can stay with the old, simple solution.

Shipit would then immediately mark the last deploy as flapping.

This would be the only way a deploy can be marked as flapping, meaning we remove the "5 times in a row" code.

@eapache @csfrancis @Sirupsen thoughts?

**Comment by Sirupsen** [comment]
With an exit-code you'd be required to `rescue` any exception and then reach out to an external data source though, which doesn't seem like the most resilient design. On the other hand, using an endpoint for the app and enforcing a certain serialized structure (that could come from something like Zookeeper) is not the most approachable either.

**Comment by Sirupsen** [comment]
This sounds good to me :+1:

**Comment by eapache** [comment]
> Non-zero exits other than 42 are ignored. This is so that application with transactional deploys (e.g. heroku apps) can stay with the old, simple solution.

Isn't the old simple solution a url not a script? I'm not sure I follow.

But generally I would prefer something a bit simpler: a script that outputs a line for each running SHA. That seems easier to manage to me?

**Comment by byroot** [comment]
> Isn't the old simple solution a url not a script? I'm not sure I follow.

No it's already a script. Even though all the usages I can think of are all like: `curl --silent https://example.com/ping/version`

>  a script that outputs a line for each running SHA

Yeah I though about that too. I think I would be ok with it. I think what made me prefer the status code, is that we need to specify the format at least a bit.

**Comment by csfrancis** [comment]
Do we care about the case when we're running a rev that wasn't deployed with Shipit? I guess Shipit can figure that out for itself if exit status 0 returns that rev. 

:+1: this sounds good to me.

**Comment by byroot** [comment]
Yup that is what I call "Manual deploy". We already support that case, and I just want to keep it supported.

**Comment by eapache** [comment]
I have a small preference for multi-line over exit code, but no good reasons for it. Either one sounds good to me.

**Comment by byroot** [comment]
Yeah I think multiline is better too.

**Comment by csfrancis** [comment]
I like exit code because it is very explicit. The issue I have with multiline output is that it seems really easy for a script to unintentionally output to stdout (either directly or through a dependency). It's not so easy to unintentionally exit with status code 42.

**Comment by byroot** [comment]
I see your point Scott. But the common case for this check is to `curl` and endpoint. and controlling exit code then is harder.

As for the output, we already discard anything that isn't a valid git SHA. So I think it would be ok. Not perfect for sure, but ok.
